parachain/pallets/inbound-queue/src/lib.rs

```diff
@@ -208,39 +231,24 @@ pub mod pallet {
 			let sovereign_account = envelope.dest.into_account_truncating();
 			T::Token::transfer(&sovereign_account, &who, T::Reward::get(), Preservation::Preserve)?;
 
-			// From this point, any errors are masked, i.e the extrinsic will
-			// succeed even if the message was not successfully decoded or dispatched.
-
-			// Attempt to decode message
+			// Decode message into XCM
 			let xcm = match inbound::VersionedMessage::decode_all(&mut envelope.payload.as_ref()) {
 				Ok(inbound::VersionedMessage::V1(message_v1)) => message_v1.into(),
 				Err(_) => {
-					Self::deposit_event(Event::MessageReceived {
-						dest: envelope.dest,
-						nonce: envelope.nonce,
-						result: MessageDispatchResult::InvalidPayload,
-					});
-					return Ok(())
-				},
+					return Err(Error::<T>::InvalidPayload.into());
+				}
 			};
 
-			// Attempt to convert to XCM
-			let sibling_para =
-				MultiLocation { parents: 1, interior: X1(Parachain(envelope.dest.into())) };
+			// Attempt to send XCM to a dest parachain
+			let dest = MultiLocation { parents: 1, interior: X1(Parachain(envelope.dest.into())) };
+			let (xcm_hash, _) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
```

Comment:
Do we still need to do any fee handling here (outside of this issue), using the send xcm price of delivery return parameter?

Reply:
Not really. I did look into it, and the fee returned by `send_xcm` (i.e the fee for sending) seems to be an optional thing that chains can decide to implement.

Ultimately, on the destination chain, some account controlled by the source chain must pay fees anyway.